src/libraries/Microsoft.PowerFx.Core/Public/Values/CollectionTableValue.cs

```diff
@@ -108,6 +114,45 @@ record = Marshal(item);
             }
         }
 
+        public override async Task<DValue<BooleanValue>> RemoveAsync(IEnumerable<FormulaValue> recordsToRemove, bool all, CancellationToken cancel)
+        {
+            var ret = false;
+
+            if (_sourceList == null)
+            {
+                return await base.RemoveAsync(recordsToRemove, all, cancel);
+            }
+
+            foreach (RecordValue recordToRemove in recordsToRemove)
+            {
+                var deleteList = new List<T>();
+
+                foreach (var current in _enumerator)
+                {
+                    cancel.ThrowIfCancellationRequested();
+
+                    var currentRecord = current as RecordValue;
```

Comment:
>r currentRecord = current as RecordValue; [](http://example.com/codeflow?start=22&length=41)

This will frequently fail. 
Marshal() instead?

Reply:
To see this, 
can you add a test to  DataTableEvalTest that uses Remove on one of those tables?
That test derives from CollectionTableValue where T is not a RecordValue, and should fail here.